### DISCUSSION TITLE
High performance C# code?

### DISCUSSION THREAD

**Discussion by Thaina** [dotnet/csharplang#4700] [Language Ideas]
Taken from unity `BurstCompile` attribute. There is an ability of unity to convert C# IL into native high performance block of code. The code in that block must be strictly use subset of C# to allow many assumption for gaining performance and using in native target. Also it could utilize concurrent and SIMD

I wish this should be available in C# too. With some keyword or attribute that C# compiler would also understand, will make a block of code safe for making native instruction, prohibit feature and library that not possible to optimized into native at compile time. And communicate with managed system with `Task` or `Observable` pattern. And, if possible, allow us to write shader and GPGPU or other low level language in C# directly

related;
- https://github.com/dotnet/vblang/issues/22
- https://github.com/dotnet/roslyn/issues/7621

**Comment by CyrusNajmabadi** [comment]
> I wish this should be available in C# too. With some keyword or attribute that C# compiler would also understand, 

If you'd like an attribute that enforces a restriction to only use a subset of C#, then that is already possible today.  That's an exact scenario that analyzers were created to solve.

**Reply by Thaina** [reply]
That was not the main point, the main point is officially support of the compiling from restricted block of C# into native target by compiler itself. The attribute and compiler restriction is just assisting feature that should be available too

**Reply by CyrusNajmabadi** [reply]
> That was not the main point, the main point is officially support of the compiling from restricted block of C# into native target by compiler itself.

this doesn't feel like the purview of the C# compiler.  It would be the job of an IL to assembler backend that could work on any appropriate IL subset that any language could emit.

**Reply by Thaina** [reply]
Didn't strict subset of language to intentionally generate proper block of IL that will be optimized in target platform if available should be start at compiler level though?

**Reply by CyrusNajmabadi** [reply]
No.  That would be the purview of an IL optimizer.  C# already generates proper blocks of IL as per  CIL.  The point of IL is that it's an *intermediate* language.  The idea of 'optimizing in target platform' would only come from an actual optimizer that knows what the target platform is and can thus adjust or limit the IL accordingly and then generate specifically for that target.

This is already what the jit does.  You're just asking for the jit to potentially run ahead of time (something which other efforts cover) or have the JIT specially optimize certain subsets differently based on knowledge it has.

**Reply by Thaina** [reply]
I think in opposie way. I think that JIT might already able to do that. But what I think is it should be a feature in C# to strictly enforce those optimization for certain subsets that written in strict language of C#, to generate IL that make assumption and constraint developer to abide by that strictness

**Reply by HaloFour** [reply]
> But what I think is it should be a feature in C# to strictly enforce those optimization for certain subsets that written in strict language of C#,

Why should C# know or care anything about a feature specific to Unity?  Any such optimization in a JIT or AOT would be very specific to that backend and would definitely evolve independently from the compiler.  You'd end up with multiple "dialects" of this "optimized" C# language for each platform and version within that platform.  That's not the job of the language.

**Reply by Thaina** [reply]
> > But what I think is it should be a feature in C# to strictly enforce those optimization for certain subsets that written in strict language of C#,
> 
> Why should C# know or care anything about a feature specific to Unity? Any such optimization in a JIT or AOT would be very specific to that backend and would definitely evolve independently from the compiler. You'd end up with multiple "dialects" of this "optimized" C# language for each platform and version within that platform. That's not the job of the language.

I don't specifically meant this feature to relate with unity. I propose a feature similar to that, just taken idea from it

C# as of now have been a cross platform language, xamarin and webgl and more. It can output to many platform but to utilize many hardware feature we then need to switch language around for interop. But most of them actually can be written in C#, as for example to use SIMD and concurrent with small instruction like GPGPU and shader, we could actually write it in subset of C# interchangeably

And I think there are not so much dialects. I propose just one for the high performance concurrent task. Which work for ECS, shader, GPGPU, or multi core processor optimization in general

**Reply by HaloFour** [reply]
So you've suggested five separate backends right there, which would be five separate dialects.  And each of those dialects would evolve separately from one another and from the compiler, leading to a Cartesian product of flavors of C# within C#.  Having to support such subsets within the compiler would invariably harm their evolution as every change would have to be coordinated across multiple teams with competing priorities and release cadences.

Describing a subset of the language or runtime available is the realm of analyzers, which are both substantially easier to write and maintain but can be done so independently from one another and evolving alongside the functionality of those backends.

It's the job of the C# compiler to produce intermediate language that is agnostic of specific backends.

**Reply by Thaina** [reply]
I don't think it would become 5 dialects. The thing that really require optimization almost overlapped. I think it would be just like PoC in C#. And other than that would be a library for specific backend

Such as mathematic and vector operation could be used in ECS and shader and GPGPU in the exact same manner. And could be optimized by JIT differently. But it should be able to run the same code

Or, for example, `return (1 + 2) / 3f;` this line should be able to run anywhere even in GPU right?

I think we could scope narrowly and allow for the code that ensure portable majorly to be optimized like this

I think this related to language design because I am not so sure we can do everything about this with just analyzer and without any language support at all. If it really possible then thank you for your clarification

**Reply by HaloFour** [reply]
> But it should be able to run the same code

In very narrow cases, maybe.  But invariably every single possible backend will have different code that it can or can't support.  And each one of those subsets will evolve over time.

> And other than that would be a library for specific backend

That library is the perfect place to ship an analyzer.  How many years do you think a backend runtime would want to wait to ship any new features because it has to wait on the compiler support to catch up?

> I think this related to language design because I am not so sure we can do everything about this with just analyzer and without any language support at all.

I disagree.  The language doesn't, and shouldn't, care about specific backend optimizations or subsets.  That is entirely the realm of analyzers and other tooling.

**Reply by CyrusNajmabadi** [reply]
> it should be a feature in C# to strictly enforce those optimization for certain subsets that written in strict language of C#, 

Sure. But as mentioned already, this capability *already* exists. This is an exact case that analyzers are for.

**Comment by huoyaoyuan** [comment]
Generally, Unity doesn't fell like a member of .NET family. It doesn't define API shape with good practice of .NET, nor process code in the .NET toolchain way. It uses old version of Mono, which could be slower in usual pathes.
